### PR TITLE
[WIP] Add units to tracker module scalarAdd API

### PR DESCRIPTION
**Pull request by mgrauer**
Multiple related PRs to add units to the tracker module's scalarAdd API method.

https://github.com/midasplatform/Midas/pull/118
https://github.com/midasplatform/pydas/pull/26

### CODE REVIEW

**Comment by mgrauer** [comment]
With the implementation on this branch, a trend with a set of params `P` will be a different row in the trend table from that same set of params intersected with a non-blank unit, `P+u`.  Meaning that an existing trend will not have continuity with that same trend's previously submitted scalars once units are included.  This may not be a problem, but is worth being aware of.

File: modules/tracker/models/base/TrendModelBase.php [hunk 1]
```diff
@@ -82,9 +82,10 @@ public function __construct()
      * @param null|int $configItemId configuration item id
      * @param null|int $testDatasetId test dataset item id
      * @param null|int $truthDatasetId truth dataset item id
+     * @param null|string $unit scalar value unit
      * @return false|Tracker_TrendDao trend DAO or false if none exists
      */
-    abstract public function getMatch($producerId, $metricName, $configItemId, $testDatasetId, $truthDatasetId);
+    abstract public function getMatch($producerId, $metricName, $configItemId, $testDatasetId, $truthDatasetId, $unit);
```
Comment (jamiesnape) [inline modules/tracker/models/base/TrendModelBase.php]: We should probably maintain the previous API, so give `$unit` a default value. I favor `false`. Do the same in any other public functions.

File: modules/tracker/models/pdo/TrendModel.php [hunk 2]
```diff
@@ -58,6 +59,10 @@ public function getMatch($producerId, $metricName, $configItemId, $testDatasetId
             $sql->where('test_dataset_id = ?', $testDatasetId);
         }
 
+        if (!is_null($unit) && ($unit !== '')) {
```
Comment (jamiesnape) [inline modules/tracker/models/pdo/TrendModel.php]: I think`WHERE unit = ''` is a valid expression.

File: modules/tracker/tests/controllers/ApiControllerTest.php [hunk 3]
```diff
@@ -75,9 +75,10 @@ public function testUploadScalarWithSubmission()
      * @param string $uuid the uuid of the submission
      * @param string $metric the metric name of the trend
      * @param float $value the scalar value
+     * @param string $unit the unit of the trend, defaults to false
```
Comment (jamiesnape) [inline modules/tracker/tests/controllers/ApiControllerTest.php]: false|string

File: modules/tracker/tests/controllers/ApiControllerTest.php [hunk 4]
```diff
@@ -75,9 +75,10 @@ public function testUploadScalarWithSubmission()
...
      * @return mixed response object from the API
      */
-    protected function _submitScalar($token, $uuid, $metric, $value)
+    protected function _submitScalar($token, $uuid, $metric, $value, $unit = false)
```
Comment (jamiesnape) [inline modules/tracker/tests/controllers/ApiControllerTest.php]: Probably `private`.

File: modules/tracker/controllers/components/ApiComponent.php [hunk 5]
```diff
@@ -243,6 +244,11 @@ public function scalarAdd($args)
         $buildResultsUrl = isset($args['buildResultsUrl']) ? $args['buildResultsUrl'] : '';
         $branch = isset($args['branch']) ? $args['branch'] : '';
 
+        if (isset($args['unit'])) {
+            $unit = $args['unit'];
+        } else {
+            $unit = null;
```
Comment (jamiesnape) [inline modules/tracker/controllers/components/ApiComponent.php]: I would prefer `false`.

File: modules/tracker/models/base/TrendModelBase.php [hunk 6]
```diff
@@ -82,9 +82,10 @@ public function __construct()
      * @param null|int $configItemId configuration item id
      * @param null|int $testDatasetId test dataset item id
      * @param null|int $truthDatasetId truth dataset item id
+     * @param null|string $unit scalar value unit
```
Comment (jamiesnape) [inline modules/tracker/models/base/TrendModelBase.php]: `false|string` if you follow my logic elsewhere.

File: modules/tracker/models/base/TrendModelBase.php [hunk 7]
```diff
@@ -136,11 +137,12 @@ public function save($trendDao)
...
```
Comment (jamiesnape) [inline modules/tracker/models/base/TrendModelBase.php]: `false|string` if you follow my logic elsewhere.

File: modules/tracker/models/pdo/TrendModel.php [hunk 8]
```diff
@@ -31,9 +31,10 @@ class Tracker_TrendModel extends Tracker_TrendModelBase
      * @param null|int $configItemId configuration item id
      * @param null|int $testDatasetId test dataset item id
      * @param null|int $truthDatasetId truth dataset item id
+     * @param null|string $unit scalar value unit
```
Comment (jamiesnape) [inline modules/tracker/models/pdo/TrendModel.php]: `false|string` if you follow my logic elsewhere.

File: modules/tracker/models/pdo/TrendModel.php [hunk 9]
```diff
@@ -31,9 +31,10 @@ class Tracker_TrendModel extends Tracker_TrendModelBase
...
      * @return false|Tracker_TrendDao trend DAO or false if none exists
      */
-    public function getMatch($producerId, $metricName, $configItemId, $testDatasetId, $truthDatasetId)
+    public function getMatch($producerId, $metricName, $configItemId, $testDatasetId, $truthDatasetId, $unit)
```
Comment (jamiesnape) [inline modules/tracker/models/pdo/TrendModel.php]: `$unit = false` if you follow my logic elsewhere.

File: modules/tracker/models/base/TrendModelBase.php [hunk 10]
```diff
@@ -136,11 +137,12 @@ public function save($trendDao)
...
      * @return Tracker_TrendDao trend DAO
      */
-    public function createIfNeeded($producerId, $metricName, $configItemId, $testDatasetId, $truthDatasetId)
+    public function createIfNeeded($producerId, $metricName, $configItemId, $testDatasetId, $truthDatasetId, $unit)
```
Comment (jamiesnape) [inline modules/tracker/models/base/TrendModelBase.php]: `$unit = false` if you follow my logic elsewhere.

File: modules/tracker/models/base/TrendModelBase.php [hunk 11]
```diff
@@ -148,7 +150,10 @@ public function createIfNeeded($producerId, $metricName, $configItemId, $testDat
             $trendDao->setProducerId($producerId);
             $trendDao->setMetricName($metricName);
             $trendDao->setDisplayName($metricName);
-            $trendDao->setUnit('');
+            if (is_null($unit)) {
```
Comment (jamiesnape) [inline modules/tracker/models/base/TrendModelBase.php]: `$unit === false` if you follow my logic elsewhere.

**Comment by jamiesnape** [comment]
First review pass completed.

File: modules/tracker/models/pdo/TrendModel.php (hunk 2 above)
Comment (mgrauer) [inline modules/tracker/models/pdo/TrendModel.php]: @jamiesnape 

I'm not sure if you are suggesting replace both parts of the test or only the empty string.

I'm changing this test to 

    if ($unit !== false) {

if that's ok with you.

Comment (jamiesnape) [inline modules/tracker/models/pdo/TrendModel.php]: Sure.

**Comment by mgrauer** [comment]
@jamiesnape QA addressed.

**Comment by jamiesnape** [comment]
LGTM. Merging.

**Pull request merged**